Add falco-pubsub integration

tembleking opened the pull request.

tembleking committed:
Add falco-pubsub integration

tembleking committed:
Add credentials data as env var

tembleking committed:
Fix error not being printed

tembleking committed:
falco-pubsub: each alert is sent to a different topic

tembleking committed:
Update functions to work with Google Pub/Sub

tembleking committed:
Adapt falco-pubsub to receive the credentials from Helm

tembleking committed:
Add support to upload Sysdig Captures to GCloud Bucket

tembleking committed:
Change image sysdig/capturer to sysdiglabs/capturer

nestorsalceda commented on falco-pubsub/main.go:
```diff
@@ -0,0 +1,151 @@
+
+}
+
+func adaptFromHelm(str string) string {
```
Please, rename to decodeFromSecret or similar. Helm is just a deployment automation tool.

tembleking replied:
Does this make more sense? I mean, this function is specifically created because Helm sends the secret in it's particular format, this is not a generic behavior.

nestorsalceda commented on falco-pubsub/main.go:
```diff
@@ -0,0 +1,151 @@
+
+	googleTopicName := subjectAndRuleSlug(alert)
+
+	topic := pubsubClient.Topic(googleTopicName)
```
Topic would be always the same.  With this approach we will need a topic for every falco rule to be subscribed.

Imagine the use case for sending **all** falco alerts to Slack and when a terminal shell is spawned in a pod, delete the pod. We want to be notified in slack about the terminal shell issue, and also we want to delete the pod automatically.

tembleking replied:
The idea behind this is that you deploy a function for a particular Falco event. When you create a function in Google Functions, you subscribe it to a Topic, the Topic is automatically created and this function gets subscribed to it. When Falco tries to send a message to a topic that doesn't work, it simply fails. We should talk about this.

nestorsalceda commented on playbooks/deploy_playbook_gke:
```diff
@@ -0,0 +1,81 @@
+
+pipenv lock --requirements | sed '/^-/ d' > requirements.txt
+
+directory="lambda"
```
Lambda is related to AWS. What do you think about calling it google-function or similar?

tembleking replied:
Sure, we can change this. What name should we pick?

nestorsalceda commented on playbooks/playbooks/__init__.py:
```diff
@@ -201,8 +217,15 @@ def _severity_from(self, priority):
 
 
 def falco_alert(event):
+    print(event)
```
Remove the print statement, please :)

tembleking replied:
My mistake. Maybe next time I should not debug with print() :joy:

nestorsalceda commented on playbooks/playbooks/__init__.py:
```diff
@@ -201,8 +217,15 @@ def _severity_from(self, priority):
+        if isinstance(data, str): # Base64 maybe?
+            try:
+                data = json.loads(base64.b64decode(data))
+            except Exception as e:
```
Catch only `binascii.Error` or `TypeError`, which are the errors expected if we are trying to decode a base64 and is incorrectly padded.

tembleking replied:
Sure, makes more sense.

nestorsalceda replied:
Cool, right now we have to import binascii and logging. Don't forget to capture also TypeError and use logging.error for errors instead of logging.info + error.

nestorsalceda commented on playbooks/playbooks/infrastructure.py:
```diff
@@ -66,25 +66,153 @@ def add_label_to_pod(self, name, label, value):
+                            security_context=client.V1SecurityContext(
+                                privileged=True
+                            ),
+                            env=[
```
Do you think that will be possible to share the major part of the body and just change the env variables passed to the PodSpec instead of copy & pasting?

tembleking replied:
Good point, we can discuss about it.

nestorsalceda commented on playbooks/specs/playbooks/start_sysdig_capture_for_container_playbook_spec.py:
```diff
@@ -8,18 +8,18 @@
 
 
-with description(playbooks.StartSysdigCaptureForContainer) as self:
+with description(playbooks.StartSysdigCaptureForContainerS3) as self:
```
I miss the test for the `StartSysdigCaptureForContainerGcloud` scenario.

tembleking replied:
You are right, we need it!

nestorsalceda commented on sysdig-capturer/docker-entrypoint.sh:
```diff
@@ -18,3 +18,9 @@ if [ -n "$AWS_ACCESS_KEY_ID" ] && [ -n "$AWS_SECRET_ACCESS_KEY" ] && [ -n "$AWS_
 fi
+
+if [ -n "$GCLOUD_BUCKET" ]; then
+
```
Remove this whitespace, please :)

nestorsalceda commented on falco-pubsub/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+COPY go.mod go.sum ./
+COPY *.go ./
+RUN GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags="-s" -o falco-pubsub
+RUN strip falco-pubsub
```
I'm loving this multistage build, but I want to keep the different images consistent so please, add a `Makefile` honoring the `docker` target.

As long as you are using go modules, other targets will be become deprecated.

tembleking replied:
I don't fully understand what you meant with keeping different images consistent. We should talk further about this.

nestorsalceda commented on playbooks/Pipfile:
```diff
@@ -15,8 +15,8 @@ playbooks = {path = "."}
-maya = "*"
-pyyaml = ">=4.2b1"
+pendulum = "*"
+pyyaml = ">=3.13"
```
We need to keep the pyyaml version to 4.2.b1 due to security issues.

tembleking replied:
When I tried to deploy it, I had an error with this version in particular, so I rolled back to a previous stable version. The new stable version is 5.1. Do I change it back to 4.2.b1 or can we move a step forward towards 5.1?

nestorsalceda commented on playbooks/playbooks/__init__.py:
```diff
@@ -201,8 +217,15 @@ def _severity_from(self, priority):
+            try:
+                data = json.loads(base64.b64decode(data))
+            except Exception as e:
+                print(e)
```
Remove the print statement, please :)

nestorsalceda reviewed (changes requested):
That's a pretty good work! There are some rough edges that we need to address before merging. All the others integrations with the KRE allows to subscribe from a topic from multiple functions (Slack, Isolate ...) and uses wildcards for routing keys.

In the case of Google Pub/Sub we can send additional data, as the `falco.<severity>.<name>` and we need an adapter in the Google Function which reads the additional data and uses it as a routing key so that we can have 2 or 3 different functions subscribed to the same topic and when a message is fanned out and received for the 3 functions, they know if they have to process the message.

On the other hand, setting up a topic for each rule is a pretty tedious task. In the other deployments `cncf` and `aws` there is an automation for creating the required resources. We need the same here, we need to automate how the topic is created and cleaned and if we need more permissions, that should be also automated.

Thanks for the awesome work @tembleking !

tembleking committed:
Changes requested to falco-pubsub PR

tembleking committed:
Multiple changes to merge falco-pubsub branch

- Add Makefile to falco-pubsub which honors the docker target
- Add a test for uploading a capture to Google Cloud
- Capture TypeError, not only binascii.Error
- Use logging.error instead of logging.info
- Whitespace removal

Signed-off-by: Federico Barcelona <fede_rico_94@hotmail.com>

tembleking committed:
Add --project to gcloud function deploy

tembleking committed:
Add Terraform manifests for pubsub

tembleking committed:
Allow wildcard rules to match

nestorsalceda commented:
Thank you very much @tembleking 

Great work!
